src/median_of_median.c

```diff
@@ -14,24 +14,65 @@ void swap(int *p, int *q){
 }
 
 int quick_select(int A[], int n, int k){
-  int i, j, pivot;
+    int i, j, pivot, eq;
 
-// 真ん中の要素をピボットとする
-  pivot = A[n/2];
-  A[n/2] = A[0];
-  A[0] = pivot;
-  for(i = j = 1; i < n; i++){
-    if(A[i] <= pivot){
-      swap(A+i, A+j);
-      j++;
+    pivot = A[n/2];
+    A[n/2] = A[0];
+    A[0] = pivot;
+    for(i = j = eq = 1; i < n; i++){
+        if(A[i] < pivot){
+            swap(A+i, A+j);
+            j++;
+        }
+        if(A[i] == pivot){
+            swap(A+i, A+eq);
+            swap(A+i, A+j);
+            j++;
+            eq++;
+        }
     }
-  }
-
-  if(j == k+1) return pivot;
-  else if(j < k+1) return quick_select(A+j, n-j, k-j);
-  else return quick_select(A+1, j-1, k);
+    
+    if((j - eq < k+1) && (k+1 <= j)) return pivot;
+    else if(j < k+1) return quick_select(A+j, n-j, k-j);
+    else return quick_select(A+eq, j-eq, k);
 }
 
+int median_of_median(int A[], int n, int k){
+    int a = n/5 + 1;
+    int B[a];
+    int i, j, pivot, eq;
+    
+    if(n <= 5) return quick_select(A,n,n/2-1);
+    else{
+        for(i = 0; i < a; i++){
+            if(i < a-1) B[i] = quick_select(A+5*i,5,2);
+            else B[a-1] = quick_select(A+5*(a-1),n-5*(a-1)+1,(n+5*(a-1))/2-1);
```

Comment:
変更してください．最後のところが特におかしいです．
```suggestion
            else B[a-1] = quick_select(A+5*(a-1),n-5*(a-1),(n-5*(a-1))/2);
```